### PR TITLE
refactor(apprunner): consolidate name validations for App Runner Resources and improve error messages

### PR DESCRIPTION
**Pull request by mazyu36**
### Issue # (if applicable)

N/A

### Reason for this change
Consolidate name validations for App Runner Resources and improve error messages.

### Description of changes
* Split regex-pattern check and length check to be more user-friendly.
  * `autoScalingConfigurationName` in `AutoScalingConfiguration`
  * `observabilityConfigurationName` in `ObservabilityConfigurationName`
  * `vpcIngressConnectionName` in `VpcIngressConnectionName`
* Add name validations.
  * `serviceName` in `ServiceName`. (Ref: [CFn document](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apprunner-service.html#cfn-apprunner-service-servicename))
  * `vpcConnectorName` in `VpcConnectorName`. (Ref: [CFn document](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apprunner-vpcconnector.html#cfn-apprunner-vpcconnector-vpcconnectorname))

### Description of how you validated changes
Modify and add unit tests.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-apprunner-alpha/lib/vpc-connector.ts [hunk 1]
```diff
@@ -136,6 +136,21 @@ export class VpcConnector extends cdk.Resource implements IVpcConnector {
...
 
+    if (props.vpcConnectorName !== undefined && !cdk.Token.isUnresolved(props.vpcConnectorName)) {
+
+      if (props.vpcConnectorName.length < 4 || props.vpcConnectorName.length > 40) {
+        throw new Error(
+          `\`vpcConnectorName\` must be between 4 and 40 characters, got: ${props.vpcConnectorName.length} characters.`,
+        );
+      }
+
+      if (!/^[A-Za-z0-9][A-Za-z0-9\-_]*$/.test(props.vpcConnectorName)) {
```
Comment (badmintoncryer) [inline packages/@aws-cdk/aws-apprunner-alpha/lib/vpc-connector.ts:147]: From [Cfn document](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apprunner-vpcconnector.html#cfn-apprunner-vpcconnector-vpcconnectorname), it seems that hyphen is not allowed for the VPC Connector name.

```
^[A-Za-z0-9][A-Za-z0-9-\\_]{3,39}$
```

**Review by badmintoncryer — CHANGES_REQUESTED** [review]
I have made just one comment! Please check it!

File: packages/@aws-cdk/aws-apprunner-alpha/lib/vpc-connector.ts (hunk 1 above)
Comment (mazyu36) [reply packages/@aws-cdk/aws-apprunner-alpha/lib/vpc-connector.ts:147]: Thanks.

I think regex pattern in Cfn document is wrong.

I successfully created a VPC Connector with a name containing hyphens.

スクリーンショット 2024-11-16 18 13 54

According to the management console, VPC connector names can only contain alphanumeric characters, hyphens, and underscores.
While the regex pattern in the CFn documentation indicates that backslashes are allowed, in practice they are not permitted.

image

I would like to provide feedback regarding this documentation to AWS team.

Comment (badmintoncryer) [reply packages/@aws-cdk/aws-apprunner-alpha/lib/vpc-connector.ts:147]: Thank you for your information! I understand.

File: packages/@aws-cdk/aws-apprunner-alpha/test/obserbability-configuration.test.ts [hunk 2]
```diff
@@ -29,15 +29,26 @@ test.each([
 test.each([
   ['tes'],
   ['test-observability-configuration-name-over-limitation'],
+])('observabilityConfigurationName length is invalide (name: %s)', (observabilityConfigurationName: string) => {
```
Comment (gracelu0) [inline packages/@aws-cdk/aws-apprunner-alpha/test/obserbability-configuration.test.ts:32]:
```suggestion
])('observabilityConfigurationName length is invalid (name: %s)', (observabilityConfigurationName: string) => {
```

File: packages/@aws-cdk/aws-apprunner-alpha/test/vpc-ingress-connection.test.ts [hunk 3]
```diff
@@ -107,3 +107,80 @@ test('create a vpcIngressConnection without a name', () => {
     },
   });
 });
+
+test.each([
+  ['tes'],
+  ['test-vpc-ingress-connection-name-over-limitation'],
+])('vpcIngressConnectionName length is invalide (name: %s)', (vpcIngressConnectionName: string) => {
```
Comment (gracelu0) [inline packages/@aws-cdk/aws-apprunner-alpha/test/vpc-ingress-connection.test.ts:114]:
```suggestion
])('vpcIngressConnectionName length is invalid (name: %s)', (vpcIngressConnectionName: string) => {
```

**Review by gracelu0 — COMMENTED** [review]
LGTM other than 2 typos, will approve once fixed :)

**Comment by mazyu36** [comment]
@gracelu0 
Thanks! I've updated.

File: packages/@aws-cdk/aws-apprunner-alpha/test/service.test.ts [hunk 4]
```diff
@@ -1781,3 +1781,42 @@ test.each([true, false])('isPubliclyAccessible is set %s', (isPubliclyAccessible
     },
   });
 });
+
+test.each([
+  ['tes'],
+  ['test-service-name-over-limitation-apprunner'],
+])('serviceName length is invalide (name: %s)', (serviceName: string) => {
```
Comment (gracelu0) [inline packages/@aws-cdk/aws-apprunner-alpha/test/service.test.ts:1788]:
```suggestion
])('serviceName length is invalid (name: %s)', (serviceName: string) => {
```

File: packages/@aws-cdk/aws-apprunner-alpha/test/vpc-connector.test.ts [hunk 5]
```diff
@@ -148,4 +148,47 @@ test('create a vpcConnector with an empty security group array should create one
...
     ],
   });
-});
\ No newline at end of file
+});
+
+test.each([
+  ['tes'],
+  ['test-vpc-connector-name-over-limitation-apprunner'],
+])('vpcConnectorName length is invalide (name: %s)', (vpcConnectorName: string) => {
```
Comment (gracelu0) [inline packages/@aws-cdk/aws-apprunner-alpha/test/vpc-connector.test.ts:156]:
```suggestion
])('vpcConnectorName length is invalid (name: %s)', (vpcConnectorName: string) => {
```

**Review by gracelu0 — COMMENTED** [review]
just two more typos sorry I missed them earlier 😅

**Comment by mazyu36** [comment]
@gracelu0 
Oops...Sorry, I've fixed again.

**Review by gracelu0 — APPROVED** [review]
Thank you!

**Pull request merged**